[Vu Van Dung] iP

joulev opened the pull request.

joulev committed:
Implement an initial skeletal version of the program (level 0)

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Rename the duke to an archduke and reformat all files

It is recommended that I rename the duke to something else. I think
archduke is a nice name for it.

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Update tests following name changes

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Set up the Java formatter in VSCode

CS2113 officially supports and recommends IntelliJ. However, I am NOT
willing to install yet another huge IDE for a single module. Being
already familiar with VSCode, I want to use it for this module as well.

This commit adds the necessary configuration files to set up the Java
formatter in VSCode, along with an editor line at 120 chars serving as
the visual limit. With this commit, my code should fully follow the
style guide without any manual intervention.

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Rename print line method to camelCase

Ref: https://se-education.org/guides/conventions/java/basic.html#naming
Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Use box-drawing characters to make the prompts look cooler

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Add a function to read user input to the IO class

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement the printLogo method for archduke

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Add vertical spacing after the boxes

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement level 1: greet, echo and exit

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement addTask and listTasks methods

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Mark all methods and props as public or private

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement marking tasks as done/undone

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Make the corners of the box more rounded

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Update README with sample output so far

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Pad index numbers with spaces in the task list

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Major refactor: standardise output processes

* Rearrange the files into packages because we just have to
* No more ugly `print*` methods inside the main Java class. They have
  now been moved to the `io.Out` class.
* Remove the check for marking/unmarking a task as done when the task
  is already done/not done. This is meant to be temporary, but I
  personally see no point in it.

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Add a parser to parse user input to machine-usable form

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Override toString() for Task class

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Finish level 4 with Todo's, Deadlines and Events

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Update tests

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Try GitHub Actions to run UI test on every push

Signed-off-by: Vu Van Dung <me@joulev.dev>

sansders commented on src/main/java/io/Out.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    private static void printBoxRightBorder(int unusedSpace) {
+        // = since we also have one space as padding
```
Comments can be clearer

sansders commented on src/main/java/Archduke.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    public static void main(String[] args) {
```
Very clean main method!

joulev replied:
Thanks!

sansders commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,15 @@
+name: UI Test
```
Can probably include this in .gitignore

joulev replied:
Actually this file is a GitHub Actions file that needs to be present in upstream. See the [documentation](https://docs.github.com/en/actions) for more info.

sansders commented on src/main/java/commands/Parser.java:
```diff
@@ -0,0 +1,51 @@
+    public static Command parse(String input) {
+        String[] options = input.split(" /");
+
+        String[] mainPart = options[0].split(" ", 2);
```
Perhaps you can change this variable name to be a plural form of a different variable name

joulev replied:
Actually I intended it to be the main part with several sections within it, so I considered plural form not necessary. On second thought now `mainPart` is not a clear variable name in the first place. Renamed.

sansders reviewed (commented):
Looks great overall! All the best for the rest of the mod!

joulev committed:
Apply changes following code review

Co-authored-by: Vu Van Dung <me@joulev.dev>
Co-authored-by: sansders <e0689644@u.nus.edu>
Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev commented:
@sansders Thanks for the reviews!

joulev committed:
Fix wrong airing dates in tests and README

Too many assignments make me lose the concept of time. Good episode
yesterday btw

Signed-off-by: Vu Van Dung <me@joulev.dev>

tyuyang commented on src/main/java/io/Out.java:
```diff
@@ -0,0 +1,153 @@
+    private static void printLogo() {
+        String left = BoxDrawingCharacter.VERTICAL_LINE + " ";
+        String right = " " + BoxDrawingCharacter.VERTICAL_LINE + "\n";
+        String[] lines = { "                   __        __      __      ",
+                "  ____ ___________/ /_  ____/ /_  __/ /_____ ", " / __ `/ ___/ ___/ __ \\/ __  / / / / //_/ _ \\",
+                "/ /_/ / /  / /__/ / / / /_/ / /_/ / ,< /  __/", "\\__,_/_/   \\___/_/ /_/\\__,_/\\__,_/_/|_|\\___/ ",
+                "                                             " };
```
Line 50 and line 51 can be split into 4 separate lines, computers with smaller aspect ratios cannot see line 51 on the same line.

joulev replied:
Actually it's due to my formatter, the line length limit is set to 120 chars as per [recommendation](https://se-education.org/guides/conventions/java/basic.html#layout) and the editor took charge of automatically add new lines where needed. I think I will reduce the length limit to 100 chars instead.

joulev replied:
Line length limit reduced to 100 chars.

tyuyang commented on src/main/java/io/Out.java:
```diff
@@ -0,0 +1,153 @@
+        printLineWithDelim(BoxDrawingCharacter.TOP_LEFT_CORNER, BoxDrawingCharacter.TOP_RIGHT_CORNER);
+    }
+
+    private static void printBoxBottomBorder() {
+        printLineWithDelim(BoxDrawingCharacter.BOTTOM_LEFT_CORNER, BoxDrawingCharacter.BOTTOM_RIGHT_CORNER);
+        System.out.println("");
+    }
+
+    private static void printBoxBottomBorder(boolean withBottomMargin) {
+        printLineWithDelim(BoxDrawingCharacter.BOTTOM_LEFT_CORNER, BoxDrawingCharacter.BOTTOM_RIGHT_CORNER);
+        if (withBottomMargin) {
+            System.out.println("");
+        }
+    }
```
Is printBoxBottomBorder() necessary? It has the same function as printBoxBottomBorder(false).

joulev replied:
Yes, the presence of both is intentional, I used overloading here to avoid having to use `printBoxBottomBorder(false)` all the time, as [Java does not allow optional parameters](https://stackoverflow.com/a/12994104).

joulev committed:
Formatter: Set maximum line length to 100 chars

Co-authored-by: Vu Van Dung <me@joulev.dev>
Co-authored-by: tyuyang <tayyuyang@gmail.com>
Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev commented:
@tyuyang Thanks

joulev committed:
Implement the ParserException class

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Set up an UnknownException error class

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Use UnknownException when the code is buggy

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Handle the case of word too long to fit in the box

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement UserInputException to handle invalid user input

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Handle case when user (un)marks with invalid index

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Update tests

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Merge branch 'branch-Level-5'

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Put all files under a dev.joulev.archduke package

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Rewrite the runtest shell script

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Install zsh in GitHub Actions runner

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Use java.util.ArrayList instead of normal arrays for tasks

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement task deletion functionality

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Set up Gradle

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Install GSON and got it working

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Set up SavedTask and Storage classes to help with fs-io

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement saving task list data to file system

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Merge branch 'branch-Level-7'

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Merge branch 'branch-Level-6'

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Change directory of saved task list file

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Install Gradle Shadow to prepare jar files

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Configure Gradle Shadow output format, bump version to 0.1

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Implement proper DateTime for the app

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Add methods to handle queries on the task list

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Add perhaps a bit too many JavaDoc comments

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Merge pull request #2 from joulev/branch-Level-9

Level 9. Find

joulev committed:
Merge pull request #1 from joulev/branch-Level-8

Level 8. Dates and Times

joulev committed:
Merge branch 'master' into branch-A-JavaDoc

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Merge pull request #3 from joulev/branch-A-JavaDoc

Add JavaDoc comments

joulev committed:
Create CNAME

joulev committed:
Update README for level 9

Signed-off-by: Vu Van Dung <me@joulev.dev>

wcwy commented on app/src/main/java/dev/joulev/archduke/Archduke.java:
```diff
@@ -0,0 +1,112 @@
+
+public class Archduke {
+    private static In in;
+    private static Store store;
```
Consider using nouns for variables. Even while both "in" and "store" has noun forms too, their meanings in noun forms do not suit well in your usage

wcwy commented on app/src/main/java/dev/joulev/archduke/commands/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     *                           or otherwise cannot be parsed.
+     */
+    public static Command parse(String input) throws ArchdukeException {
+        String[] options = input.split(" /");
```
Magic literals here, consider using named constant for " /".

wcwy replied:
There are other similar magic literals issues found in other files too. Be careful on this issue.

wcwy commented on app/src/main/java/dev/joulev/archduke/commands/Parser.java:
```diff
@@ -0,0 +1,75 @@
+
+        // Separating the main command from the parameters aka the body
+        String[] components = options[0].split(" ", 2);
+        String type = components[0];
```
What "type" are you referring to? Might want to make this clear in the variable naming itself.

wcwy commented on app/src/main/java/dev/joulev/archduke/io/In.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    public String readUserInput() {
+        if (System.getenv("TESTING") == null) {
+            System.out.print("> ");
```
Magic string

wcwy commented on app/src/main/java/dev/joulev/archduke/commands/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     * @throws ArchdukeException If the input is invalid, contains unknown option(s)
+     *                           or otherwise cannot be parsed.
+     */
+    public static Command parse(String input) throws ArchdukeException {
```
The whole parse method is quite long. You might want to refactor the parsing logic into few methods and apply SLAP here to make something similar to this:
parse() {
&nbsp;&nbsp;&nbsp;&nbsp;splitCommandTypeAndParameters();
&nbsp;&nbsp;&nbsp;&nbsp;parseParameters();
&nbsp;&nbsp;&nbsp;&nbsp;prepareCommand();
}

wcwy commented on app/src/main/java/dev/joulev/archduke/Archduke.java:
```diff
@@ -0,0 +1,112 @@
+                case "find":
+                    Out.printQueriedTasks(store, command.getBody());
+                    continue;
+                case "mark":
```
Missing fallthrough comment here. Take a look at Java Coding Standard document on this!

wcwy commented on app/src/main/java/dev/joulev/archduke/storage/Storage.java:
```diff
@@ -0,0 +1,84 @@
+public class Storage {
+    /** The path to the JSON file storing the task list. */
+    private static final String FILE = ".archduke/tasks.json";
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
```
Should `gson` be in full caps too?

wcwy commented on app/src/main/java/dev/joulev/archduke/storage/Storage.java:
```diff
@@ -0,0 +1,84 @@
+            // Whether the file is not found, is corrupted or for whatsoever reason, we
+            // consider it as non-existent and return an empty list.
+            return new ArrayList<>();
+        }
```
Does this mean that even if only one particular entry is corrupted, then the whole file will be wiped out if the user performs any further function that triggers `writeToFile()`? If that's the case, you might want to consider if there is better way to approach this. 

Also, you might want to consider giving user a message to alert the user on their file issues too.

wcwy reviewed (commented):
Most of your code follows the coding standards and maintains a good coding quality. Good job there! Do look out for naming convention, magic numbers and other commented issues below.

joulev committed:
Rename classes and variables from review comments

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Rename more variables and classes

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Refactor the parser to make it more understandable

hmm, I definitely see this as just more verbose and not a slightest
more understandable, but...

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Make the last final variable in THIS_CASE

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Fallthrough

Yes I know why they use fallthrough comments in normal switch
statements. But this one there's nothing under `case "mark"`, anyone
knowing anything about the programming language should already know the
thing is fallen through. A comment like this will only make Elon happy.

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Show error message if stored file is corrupted

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Squash many magic literals in printing logic and elsewhere

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev commented:
Thanks.

joulev committed:
Add user guide

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Add some SEO tags and modifications to Jekyll layout

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Shamelessly put my name in contributors.md

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev committed:
Ensure no error messages are printed if tasks.json is not found

Signed-off-by: Vu Van Dung <me@joulev.dev>

joulev commented:
Closing to partially clear my [pull request](https://github.com/pulls) list.
